src/main/java/org/opensearch/flowframework/util/JsonToJsonTransformer.java

```diff
@@ -0,0 +1,384 @@
+ *   JSON provider instead of json-smart to avoid dependency conflicts.
+ * - Only [*] wildcards are supported (nested wildcards are
+ *   allowed). Explicit numeric indices (e.g. [0]) are honoured
+ *   verbatim.
+ * - Thread-safe and stateless: all public APIs are static and use local
+ *   variables only.
+ *
+ * Call Flow Architecture:
+ * transform()
+ *   └── buildOutputFromMapping() [validates object type and processes properties]
+ *       ├── processSimpleProperty() [handles JsonPath leaf nodes]
+ *       ├── [nested objects] → buildOutputFromMapping() [direct recursion]
+ *       └── processArrayProperty() [handles array syntax and builds elements]
+ *           └── buildOutputFromMapping() [direct recursion for array elements]
+ */
+public final class JsonToJsonTransformer {
+    private static final ObjectMapper MAPPER;
+    private static final Configuration JSON_PATH_CFG;
+
+    // Pre-compiled regex patterns for performance
+    private static final Pattern EXPLICIT_INDEX_PATTERN = Pattern.compile("\\[\\d+\\]");
+    private static final Pattern WILDCARD_PATTERN = Pattern.compile("\\[\\*\\]");
+
+    static {
+        StreamReadConstraints constraints = StreamReadConstraints.builder()
+            .maxNestingDepth(CommonValue.MAX_JSON_NESTING_DEPTH)
+            .maxStringLength(CommonValue.MAX_JSON_SIZE)
+            .maxNameLength(CommonValue.MAX_JSON_NAME_LENGTH)
+            .build();
+
+        MAPPER = new ObjectMapper();
+        MAPPER.getFactory().setStreamReadConstraints(constraints);
+
+        JSON_PATH_CFG = Configuration.builder()
+            .jsonProvider(new JacksonJsonProvider(MAPPER))
+            .mappingProvider(new JacksonMappingProvider(MAPPER))
+            .options(Option.ALWAYS_RETURN_LIST, Option.SUPPRESS_EXCEPTIONS)
+            .build();
+    }
+
+    /** Utility class – no instantiation. */
+    private JsonToJsonTransformer() {}
+
+    /**
+     * Transforms {@code inputJson} into a new JSON document according to the
+     * mapping defined in {@code mappingRules}.
+     *
+     * @param inputJson   source document (string)
+     * @param mappingRules mapping produced by JsonToJsonRecommender
+     * @return transformed JSON as a string
+     * @throws JsonProcessingException if either input cannot be parsed or the
+     *                                 mapping is malformed
+     */
+    public static String transform(String inputJson, String mappingRules) throws IllegalArgumentException, JsonProcessingException {
+        if (Strings.isNullOrEmpty(inputJson)) {
+            throw new IllegalArgumentException("inputJson must not be null or empty");
+        }
+        if (Strings.isNullOrEmpty(mappingRules)) {
+            throw new IllegalArgumentException("mappingRules must not be null or empty");
+        }
+
+        JsonNode mappingNode = MAPPER.readTree(mappingRules);
+
+        Object document;
+        try {
+            document = JsonPath.using(JSON_PATH_CFG).parse(inputJson).json();
```

Comment:
The input json is already a string, we can use .jsonString() which returns a string rather than .json() which is not strongly typed.